### ISSUE TITLE
Problems building against Lua 5.3

### ISSUE THREAD

**Issue by Darksonn** [awesomeWM/awesome#361]
I am building my awesome installation from source, and I decided to pull to update, and then do as posted in [this comment in another issue](https://github.com/awesomeWM/awesome/issues/359#issuecomment-125310961). Note that the following compilation error happens regardless of the change in the other issue.

    CMakeFiles/awesome.dir/dbus.c.o: In function `a_dbus_message_iter':
    /home/user/src/awesome/.build-archan-x86_64-unknown-linux-gnu-5.2.0/dbus.c:120: undefined reference to `lua_rotate'
    /home/user/src/awesome/.build-archan-x86_64-unknown-linux-gnu-5.2.0/dbus.c:194: undefined reference to `lua_rotate'
    /home/user/src/awesome/.build-archan-x86_64-unknown-linux-gnu-5.2.0/dbus.c:211: undefined reference to `lua_rotate'
    CMakeFiles/awesome.dir/dbus.c.o: In function `luaA_object_push':
    /home/user/src/awesome/.build-archan-x86_64-unknown-linux-gnu-5.2.0/common/luaobject.h:148: undefined reference to `lua_rotate'
    CMakeFiles/awesome.dir/dbus.c.o: In function `luaA_dofunction':
    /home/user/src/awesome/.build-archan-x86_64-unknown-linux-gnu-5.2.0/common/lualib.h:69: undefined reference to `lua_rotate'
    CMakeFiles/awesome.dir/dbus.c.o:/home/user/src/awesome/.build-archan-x86_64-unknown-linux-gnu-5.2.0/common/lualib.h:73: more undefined references to `lua_rotate' follow
    collect2: error: ld returned 1 exit status
    CMakeFiles/awesome.dir/build.make:1064: recipe for target 'awesome' failed
    make[3]: *** [awesome] Error 1
    CMakeFiles/Makefile2:61: recipe for target 'CMakeFiles/awesome.dir/all' failed
    make[2]: *** [CMakeFiles/awesome.dir/all] Error 2
    Makefile:116: recipe for target 'all' failed
    make[1]: *** [all] Error 2
    Makefile:27: recipe for target 'cmake-build' failed
    make: *** [cmake-build] Error 2

The makefile also seems to claim that lua 5.3.1 is being used even though it's using `/usr/lib64/liblua5.2.so`. I have confirmed that the given so file is actually the file from the pacman lib52 package.

    Found Lua: /usr/lib64/liblua5.2.so;/usr/lib64/libm.so (found version "5.3.1")

I can't remember whether awesome needed lua51 or lua52 and it's not apparent from the README, but both are installed on my system. I assume that if I really needed lua51 to compile, it worked last time I compiled since I didn't have the other version of lua installed at that time. Now however I've installed other programs that depend on lua52, if this is the issue, how can I make awesome use the correct version of lua?

**Comment by blueyed** [comment]
Awesome from master should be fine with both 5.2 and 5.3, but not when it gets mixed up.

What version of CMake are you using?
Maybe your `FindLua` is buggy?  It gets used here: https://github.com/awesomeWM/awesome/blob/b505aa1a5d4e9a3339cdcd45759fd62b329f2693/awesomeConfig.cmake#L62

**Comment by blueyed** [comment]
If you can, just try it with only Lua 5.2 or 5.3 being installed - or only 5.1 and 5.2 (since your 5.3 appears to be problematic).

**Comment by Darksonn** [comment]
I removed the lua51 package, that did not resolve the issue. I have never installed lua 5.3.

The only version of lua currently installed on my system is lua 5.2.4-1

[Here is the full output of make after removing lua5.1](https://gist.github.com/Darksonn/c82c62b63595d88e99be)

It still contains the line

    Found Lua: /usr/lib64/liblua5.2.so;/usr/lib64/libm.so (found version "5.3.1")

**Comment by blueyed** [comment]
What's you CMake version?  Can you update it?
What distro are you using?

The problem is clearly that somehow somewhere `5.3.1` gets found.
Are you sure there's nothing in /usr/local etc?

You need to dig into what CMake is doing, and where it finds 5.3.1.

**Comment by Darksonn** [comment]
Okay it appears I was wrong, the lua 5.3 package was simply not following the same naming pattern that the other packages were using.

Removing lua 5.2 fixes the compilation issue. Reinstalling it causes the issue again.

After looking at the FindLua code, I'm pretty sure it's because it tries to find the lua version by looking at `/usr/include/lua.h`. I only have one of those (find says so) in my `/usr` directory and the file contains the codes for lua 5.3, however pacman claims that both lua52 and lua53 owns the file. That's probably the issue.

I'm using CMake version 3.2.3-1, it's the newest version in the package repository, and appears to be the newest.

The distro is arch linux.

**Comment by blueyed** [comment]
I've installed Lua 5.3 manually myself now and this triggered the same problem.

Lua installs `/usr/local/lib/liblua.a`, which does not get picked up by CMake
apparently, but it uses the .so from 5.2 instead.

For awesome to build against 5.3, I had to set:

    LUA_LIBRARY=/usr/local/lib/liblua.a;dl

This can be done via e.g. `cmake-gui`, or in the Makefile:

    CMAKE_ARGS:="-DLUA_LIBRARY=/usr/local/lib/liblua.a\;dl"

Note the additional `dl`, without this linking fails:

    Linking C executable awesome
    /usr/bin/ld: /usr/local/lib/liblua.a(loadlib.o): undefined reference to symbol 'dlclose@@GLIBC_2.2.5'
    /lib/x86_64-linux-gnu/libdl.so.2: error adding symbols: DSO missing from command line
    collect2: error: ld returned 1 exit status


I am using cmake 3.0.2-1ubuntu2.

**Comment by Bew78LesellB** [comment]
I had this problem too, I used `CMAKE_ARGS:="-DLUA_LIBRARY=/usr/lib/liblua.so\;dl"` to make it work if this can help

**Comment by psychon** [comment]
This is a CMake bug, isn't it? (And a side-effect of Lua having no pkg-config file and no "real" standard way for installing)

Google found https://cmake.org/Bug/view.php?id=15756.

I am closing this issue under the assumption that the bug is in CMake. Feel free to reopen if anyone disagrees.
